src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -71,9 +73,9 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
                 var command = CreateStartCommand(currentStep);
 
                 var speculativeAuthenticateResult = description.IsMasterResult.SpeculativeAuthenticate;
-                if (speculativeAuthenticateResult != null)
+                if (speculativeAuthenticateResult != null && _speculativeFirstStep != null)
                 {
-                    currentStep = Transition(conversation, currentStep, speculativeAuthenticateResult, out command);
+                    currentStep = Transition(conversation, _speculativeFirstStep, speculativeAuthenticateResult, out command);
```

Comment:
Note: lines 72-73 are only needed when there is no speculative auth result.

Refactor lines 72-79 to:

    ISaslStep currentStep;
    BsonDocument command;
    var speculativeAuthenticateResult = description.IsMasterResult.SpeculativeAuthenticate;
    if (_speculativeFirstStep != null && speculativeAuthenticateResult != null)
    {
        currentStep = Transition(conversation, _speculativeFirstStep, speculativeAuthenticateResult, out command);
    }
    else
    {
        currentStep = _mechanism.Initialize(connection, conversation, description);
        command = CreateStartCommand(currentStep);
    }

Reply:
Took your refactor!